File: textdb/textdb-api/src/main/java/edu/uci/ics/textdb/api/tuple/Tuple.java
package edu.uci.ics.textdb.api.tuple;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;

import edu.uci.ics.textdb.api.constants.JsonConstants;
import edu.uci.ics.textdb.api.exception.TextDBException;
import edu.uci.ics.textdb.api.field.IField;
import edu.uci.ics.textdb.api.schema.Schema;

/**
 * @author chenli
 * @author sandeepreddy602
 * @author zuozhi
 * 
 * Created on 3/25/16.
 */
@JsonDeserialize(using = TupleJsonDeserializer.class)
public class Tuple {
    private final Schema schema;
    private final List<IField> fields;

    public Tuple(Schema schema, IField... fields) {
        this(schema, Arrays.asList(fields));
    }
    
    @JsonCreator
    public Tuple(
            @JsonProperty(value = JsonConstants.SCHEMA, required = true)
            Schema schema, 
            @JsonProperty(value = JsonConstants.FIELDS, required = true)
            List<IField> fields) {
        validateTuple(schema, fields);
        this.schema = schema;
        this.fields = fields;
    }
    
    @JsonProperty(value = JsonConstants.SCHEMA)
    public Schema getSchema() {
        return schema;
    }
    
    @JsonProperty(value = JsonConstants.FIELDS)
    public List<IField> getFields() {
        return new ArrayList<>(this.fields);
    }
    
    @SuppressWarnings("unchecked")
    public <T extends IField> T getField(int index) {
        if (index < fields.size()) {
            return (T) fields.get(index);
        }
        throw new TextDBException(String.format(
                "Cannot get field at index %d: tuple only has %d fields", 
                index, fields.size()));
    }
    
    public <T extends IField> T getField(int index, Class<T> fieldClass) {
        return getField(index);
    }

    public <T extends IField> T getField(String attributeName) {
        return getField(schema.getIndex(attributeName));
    }
    
    public <T extends IField> T getField(String attributeName, Class<T> fieldClass) {
        return getField(schema.getIndex(attributeName));
    }
    
    /*
     * Validates if the schema and the fields are consistent. 
     * Throws TextDBException if not.
     */
    private void validateTuple(Schema schema, List<IField> fields) throws TextDBException {
        // check if the size of schema and fields are the same
        if (schema.size() != fields.size()) {
            throw new TextDBException(String.format(
                    "Cannot create Tuple: schema has %d attributes, but there are %d fields.", 
                    schema.size(),
                    fields.size()));
        }
        // check if the field classes are consistent with the schema's attribute types
        for (int i = 0; i < fields.size(); i++) {
            if (! schema.getAttribute(i).getAttributeType().getFieldClass().equals(
                    fields.get(i).getClass())) {
                throw new TextDBException(String.format(
                        "Cannot create Tuple: attribute %s has type %s, but got %s",
                        schema.getAttribute(i).getAttributeName(), schema.getAttribute(i).getAttributeType(),
                        IField.fieldTypeMap.get(fields.get(i).getClass())));
            }
        }
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + ((fields == null) ? 0 : fields.hashCode());
        result = prime * result + ((schema == null) ? 0 : schema.hashCode());
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj == null)
            return false;
        if (getClass() != obj.getClass())
            return false;
        Tuple other = (Tuple) obj;
        if (fields == null) {
            if (other.fields != null)
                return false;
        } else if (!fields.equals(other.fields))
            return false;
        if (schema == null) {
            if (other.schema != null)
                return false;
        } else if (!schema.equals(other.schema))
            return false;
        return true;
    }

    @Override
    public String toString() {
        return "Tuple [schema=" + schema + ", fields=" + fields + "]";
    }

}

File: core/amber/src/main/scala/edu/uci/ics/texera/workflow/common/tuple/Tuple.scala
package edu.uci.ics.texera.workflow.common.tuple

import com.fasterxml.jackson.annotation.{JsonCreator, JsonIgnore, JsonProperty}
import com.fasterxml.jackson.databind.node.ObjectNode
import com.fasterxml.jackson.databind.JsonNode
import edu.uci.ics.texera.workflow.common.tuple.schema.{Attribute, AttributeType, Schema}
import edu.uci.ics.amber.engine.common.tuple.amber.SeqTupleLike
import edu.uci.ics.texera.Utils
import edu.uci.ics.texera.workflow.common.tuple.exception.TupleBuildingException
import org.bson.Document
import org.ehcache.sizeof.SizeOf
import scala.collection.mutable
import scala.util.hashing.MurmurHash3

case class Tuple @JsonCreator() (
    @JsonProperty(value = "schema", required = true) schema: Schema,
    @JsonProperty(value = "fields", required = true) fieldVals: List[Any]
) extends SeqTupleLike
    with Serializable {

  import Tuple._

  require(schema != null, "Schema cannot be null")
  require(fieldVals != null, "Fields cannot be null")
  checkSchemaMatchesFields(schema.getAttributes, fieldVals)

  override val inMemSize: Long = SizeOf.newInstance().deepSizeOf(this)

  @JsonIgnore def length: Int = fieldVals.size

  @JsonIgnore def get(i: Int): Any = fieldVals(i)

  @JsonIgnore def getSchema: Schema = schema

  def getField[T](attributeName: String): T = {
    if (!schema.containsAttribute(attributeName)) {
      throw new RuntimeException(s"$attributeName is not in the tuple")
    }
    fieldVals(schema.getIndex(attributeName)).asInstanceOf[T]
  }

  def getField[T](attributeName: String, fieldClass: Class[T]): T = getField[T](attributeName)

  def getFields: Seq[Any] = fieldVals

  override def hashCode: Int = util.Arrays.deepHashCode(fields.map(_.asInstanceOf[AnyRef]))

  override def equals(obj: Any): Boolean =
    obj match {
      case that: Tuple => (this.fields sameElements that.fields) && this.schema == that.schema
      case _           => false
    }

  def getPartialTuple(indices: Array[Int]): Tuple = {
    val partialSchema = schema.getPartialSchema(indices)
    val builder = Tuple.BuilderV2(partialSchema)
    val partialArray = indices.map(fieldVals(_))
    builder.addSequentially(partialArray)
    builder.build()
  }

  override def toString: String = s"Tuple [schema=$schema, fields=$fieldVals]"

  def asKeyValuePairJson(): ObjectNode = {
    val objectNode = Utils.objectMapper.createObjectNode()
    this.schema.getAttributeNames.foreach { attrName =>
      val valueNode = Utils.objectMapper.convertValue(this.getField(attrName), classOf[JsonNode])
      objectNode.set[ObjectNode](attrName, valueNode)
    }
    objectNode
  }

  def asDocument(): Document = {
    val doc = new Document()
    this.schema.getAttributeNames.foreach { attrName =>
      doc.put(attrName, this.getField(attrName))
    }
    doc
  }

  override def fields: Array[Any] = fieldVals.toArray
}

object Tuple {

  private def checkSchemaMatchesFields(
      attributes: Iterable[Attribute],
      fields: Iterable[Any]
  ): Unit = {
    val attributeList = attributes.toList
    val fieldList = fields.toList

    if (attributeList.size != fieldList.size) {
      throw new RuntimeException(
        s"Schema size (${attributeList.size}) and field size (${fieldList.size}) are different"
      )
    }

    (attributeList zip fieldList).foreach {
      case (attribute, field) =>
        checkAttributeMatchesField(attribute, field)
    }
  }

  private def checkAttributeMatchesField(attribute: Attribute, field: Any): Unit = {
    if (
      field != null && attribute.getType != AttributeType.ANY && !field.getClass.equals(
        attribute.getType.getFieldClass
      )
    ) {
      throw new RuntimeException(
        s"Attribute ${attribute.getName}'s type (${attribute.getType}) is different from field's type (${AttributeType
          .getAttributeType(field.getClass)})"
      )
    }
  }

  def newBuilder(schema: Schema): BuilderV2 = {
    Tuple.BuilderV2(schema)
  }

  case class BuilderV2(schema: Schema) {
    private val fieldNameMap = mutable.Map.empty[String, Any]

    def add(tuple: Tuple, isStrictSchemaMatch: Boolean = true): BuilderV2 = {
      require(tuple != null, "Tuple cannot be null")

      tuple.fields.zipWithIndex.foreach {
        case (field, i) =>
          val attribute = tuple.schema.getAttributes(i)
          if (!isStrictSchemaMatch && !schema.containsAttribute(attribute.getName)) {
            // Skip if not matching in non-strict mode
          } else {
            add(attribute, tuple.fields(i))
          }
      }
      this
    }

    def add(attribute: Attribute, field: Any): BuilderV2 = {
      require(attribute != null, "Attribute cannot be null")
      checkAttributeMatchesField(attribute, field)

      if (!schema.containsAttribute(attribute.getName)) {
        throw new TupleBuildingException(
          s"${attribute.getName} doesn't exist in the expected schema."
        )
      }

      fieldNameMap.put(attribute.getName.toLowerCase, field)
      this
    }

    def add(attributeName: String, attributeType: AttributeType, field: Any): BuilderV2 = {
      require(
        attributeName != null && attributeType != null,
        "Attribute name and type cannot be null"
      )
      this.add(new Attribute(attributeName, attributeType), field)
      this
    }

    def addSequentially(fields: Array[Any]): BuilderV2 = {
      require(fields != null, "Fields cannot be null")
      checkSchemaMatchesFields(schema.getAttributes, fields)
      schema.getAttributes.zip(fields).foreach {
        case (attribute, field) =>
          this.add(attribute, field)
      }
      this
    }

    def build(): Tuple = {
      val missingAttributes =
        schema.getAttributes.filterNot(attr => fieldNameMap.contains(attr.getName.toLowerCase))
      if (missingAttributes.nonEmpty) {
        throw new TupleBuildingException(
          s"Tuple does not have the same number of attributes as schema. Missing attributes are $missingAttributes"
        )
      }

      val fields =
        schema.getAttributes.map(attr => fieldNameMap(attr.getName.toLowerCase))
      new Tuple(schema, fields)
    }
  }
}
